File: graph/number_of_operations_to_make_the_network_connected.java
class Solution {
    public void dfs(int node, List<List<Integer>> adj, int[] visited){
        visited[node] = 1;
        for(int it : adj.get(node)){
            if(visited[it] == 0){
                dfs(it, adj, visited);
            }
        }
    }
    public int makeConnected(int n, int[][] connections) {
        //creating an adjencancy list from the connections 2D array
        List<List<Integer>> adj = new ArrayList<>();
        int num = connections.length;
        if(n - 1 > num) return -1;
        for(int i = 0 ; i < n ; i++){
            adj.add(new ArrayList<>());
        }
        for(int[] connection : connections){
            int u = connection[0];
            int v = connection[1];
            adj.get(u).add(v);
            adj.get(v).add(u);
        }
        int[] vis = new int[n];
        int ans = 0;
        for(int i = 0 ; i < n ; i++){
            if(vis[i] == 0){
                ans++;
                dfs(i, adj, vis);
            }
        }
        return ans-1;
    }
}

File: number_of_operations_to_make_the_network_connected.cpp
class Solution {
public:
    void dfs(vector<int> adj[], vector<bool> &visited, int curr){
        visited[curr] = true;
        for(auto it : adj[curr]){
            if(visited[it] == false){
                dfs(adj, visited, it);
            }
        }
    }
    int makeConnected(int n, vector<vector<int>>& connections) {
        vector<int> adj[n];
        vector<bool> visited(n , false);
        int edge = 0;
        for(auto v : connections){
            adj[v[0]].push_back(v[1]);
            adj[v[1]].push_back(v[0]);
            edge++;
        }

        if(edge < n-1) return -1;

        int components=0;
        for(int i = 0 ; i < n ; i++){
            if(visited[i] == false){
                components+=1;
                dfs(adj, visited, i);
            }
        }
        int red = edge - (n-1 - (components - 1));
        if(red >= components-1){
            return components-1;
        }
        return -1;
    }
};

